File: app/src/main/java/de/davis/passwordmanager/filter/Filter.java
package de.davis.passwordmanager.filter;

import com.google.android.material.chip.ChipGroup;

import java.util.ArrayList;
import java.util.List;

import de.davis.passwordmanager.R;
import de.davis.passwordmanager.database.ElementType;
import de.davis.passwordmanager.database.dtos.SecureElement;
import de.davis.passwordmanager.database.entities.details.password.PasswordDetails;
import de.davis.passwordmanager.database.entities.details.password.Strength;

public class Filter {

    public static final Filter DEFAULT = new Filter();

    private static final int ID_PASSWORD = R.id.password;
    private static final int ID_CREDIT_CARD = R.id.creditCard;

    private static final int ID_VERY_STRONG = R.id.veryStrong;
    private static final int ID_STRONG = R.id.strong;
    private static final int ID_MODERATE = R.id.moderate;
    private static final int ID_WEAK = R.id.weak;
    private static final int ID_RIDICULOUS = R.id.ridiculous;

    private ChipGroup type;
    private ChipGroup strength;

    private final List<Integer> selectedIds = new ArrayList<>();

    private Filter(){}

    public void setType(ChipGroup type) {
        this.type = type;
    }

    public void setStrength(ChipGroup strength) {
        this.strength = strength;
    }

    public Runnable updater;

    public void setUpdater(Runnable updater) {
        this.updater = updater;
    }

    public void update() {
        if(updater != null)
            updater.run();

        if(!groupsSet())
            return;

        selectedIds.clear();
        selectedIds.addAll(type.getCheckedChipIds());
        selectedIds.addAll(strength.getCheckedChipIds());
    }

    public List<Integer> getSelectedIds() {
        return selectedIds;
    }

    private boolean groupsSet(){
        return type != null && strength != null;
    }

    public List<SecureElement> filter(List<SecureElement> elements) {
        if(!groupsSet())
            return elements;

        List<SecureElement> toFilter = new ArrayList<>(elements);
        List<Integer> typeIds = type.getCheckedChipIds();
        List<Integer> strengthIds = strength.getCheckedChipIds();
        if(!typeIds.contains(ID_CREDIT_CARD))
            toFilter.removeIf(element -> element.getElementType() == ElementType.CREDIT_CARD);


        if(!typeIds.contains(ID_PASSWORD)) {
            toFilter.removeIf(element -> element.getElementType() == ElementType.PASSWORD);
            return toFilter;
        }

        toFilter.removeIf(element -> {
            if(element.getElementType() != ElementType.PASSWORD)
                return false;

            return !strengthIds.contains(ID_VERY_STRONG) && ((PasswordDetails)element.getDetail()).getStrength() == Strength.VERY_STRONG
                    || !strengthIds.contains(ID_STRONG) && ((PasswordDetails)element.getDetail()).getStrength() == Strength.STRONG
                    || !strengthIds.contains(ID_MODERATE) && ((PasswordDetails)element.getDetail()).getStrength() == Strength.MODERATE
                    || !strengthIds.contains(ID_WEAK) && ((PasswordDetails)element.getDetail()).getStrength() == Strength.WEAK
                    || !strengthIds.contains(ID_RIDICULOUS) && ((PasswordDetails)element.getDetail()).getStrength() == Strength.RIDICULOUS;
        });

        return toFilter;
    }
}

File: app/src/main/java/de/davis/passwordmanager/filter/Filter.kt
package de.davis.passwordmanager.filter

import de.davis.passwordmanager.database.ElementType
import de.davis.passwordmanager.database.dtos.SecureElement
import de.davis.passwordmanager.database.entities.details.password.PasswordDetails
import de.davis.passwordmanager.database.entities.details.password.Strength
import de.davis.passwordmanager.database.entities.onlyCustoms
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow

object Filter {

    private val _filterFlow = MutableStateFlow(FilterOptions())
    val filterFlow = _filterFlow.asStateFlow()

    private val _tagsFlow = MutableStateFlow(listOf<String>())
    val tagsFlow = _tagsFlow.asStateFlow()

    fun updateFilter(block: FilterOptions.() -> Unit) {
        val current = filterFlow.value
        val filter = current
            .copy(tags = mutableListOf<String>().apply { addAll(current.tags) })
            .apply(block)
        _filterFlow.value = filter
    }

    fun updateTags(tags: List<String>) {
        _tagsFlow.value = tags
    }

    data class FilterOptions(
        var password: StrengthFilter? = StrengthFilter(),
        var creditCard: Boolean = true,
        var tags: MutableList<String> = mutableListOf() //empty = all
    )

    data class StrengthFilter(
        var veryStrong: Boolean = true,
        var strong: Boolean = true,
        var moderate: Boolean = true,
        var weak: Boolean = true,
        var ridiculous: Boolean = true
    )
}

fun Collection<SecureElement>.applyFilter(filter: Filter.FilterOptions): List<SecureElement> {
    return filter { if (it.elementType == ElementType.CREDIT_CARD) filter.creditCard else true }
        .filter { if (it.elementType == ElementType.PASSWORD) filter.password != null else true }
        .filter {
            if (it.elementType != ElementType.PASSWORD)
                return@filter true

            (it.detail as PasswordDetails).let { pwdDetail ->
                filter.password?.let { pwd ->
                    pwd.veryStrong && pwdDetail.strength == Strength.VERY_STRONG
                            || pwd.strong && pwdDetail.strength == Strength.STRONG
                            || pwd.moderate && pwdDetail.strength == Strength.MODERATE
                            || pwd.weak && pwdDetail.strength == Strength.WEAK
                            || pwd.ridiculous && pwdDetail.strength == Strength.RIDICULOUS
                } ?: true
            }
        }
        .filter {
            if (filter.tags.isEmpty()) true else filter.tags.any { tag ->
                it.tags.onlyCustoms().map { it.name }.contains(tag)
            }
        }
}